base/shared/src/main/scala/scalaz/data/Void.scala

```diff
@@ -1,13 +1,45 @@
 package scalaz
 package data
 
+import com.github.ghik.silencer.silent
+
+trait VoidModule {
+  type Void
+
+  def absurd[A](v: Void): A
+
+  def unsafeVoid: Void
+
+  def isNothing: Void === Nothing
+
+  def conforms[A]: Void <~< A
+}
+
 trait VoidFunctions {
-  def unsafeVoid: Void = throw new Void.Absurd
+  @inline final def absurd[A](v: Void): A = Void.absurd[A](v)
 }
 
-object Void extends VoidFunctions {
-  private[data] trait Tag
-  type Void <: Nothing with Tag
+trait VoidSyntax {
+  implicit class Ops(v: Void) {
+    def absurd[A]: A = Void.absurd(v)
+  }
+}
+
+// NOTE: this is some next level black compiler magic
+// but without this object syntax doesn't resolve...
+object VoidSyntax extends VoidSyntax
```

Comment:
I'm curious: what's the actual compiler magic?

Reply:
I am honestly not sure *how* and *why* it works. Took it from @TomasMikula 's `Forall` implementation, maybe he knows :)

Reply:
Yeah, it's super confusing...

We are looking for implicits related to `scalaz.data.Void.Void` = `scalaz.data.Void.type#Void`.

The spec [says](https://scala-lang.org/files/archive/spec/2.12/07-implicits.html#implicit-parameters) that the "parts" of this type projection include the parts of `scalaz.data.Void.type`:

 > if T is a type projection S#U, the parts of S as well as T itself;

Then the spec says that the parts of this singleton type `scalaz.data.Void.type` are the parts of the type of `scalaz.data.Void`:

> if T is a singleton type p.type, the parts of the type of p;

The type of `scalaz.data.Void` is `VoidModule with VoidSyntax`.

The parts of `VoidModule with VoidSyntax` is just `VoidModule with VoidSyntax` itself:

> in all other cases, just T itself.

From that we conclude that `VoidModule with VoidSyntax` is a "part" of the type `scalaz.data.Void.Void`.

The class `VoidSyntax` is a base class of `VoidModule with VoidSyntax`, which is a part of `scalaz.data.Void.Void`, thus `VoidSyntax` is "associated" with `scalaz.data.Void.Void`:

> Here, we say a class C is associated with a type T if it is a base class of some part of T.

Finally,

> The implicit scope of a type T consists of all companion modules of classes that are associated with the implicit parameter's type.

which means that the companion of  `VoidSyntax` is included in the search.

Reply:
Ah! that makes sense. I knew about the prefix rules but I didn't notice that `VoidModule` extended `VoidSyntax`.

Reply:
It doesn't :). But we could have

```scala
object VoidModule extends VoidSyntax
```

instead of

```scala
object VoidSyntax extends VoidSyntax
```

and then

```scala
val Void: VoidModule = VoidImpl
```

instead of

```scala
val Void: VoidModule with VoidSyntax = VoidImpl
```

I guess that'd be slightly better.